vms/proposervm/block/block.go

```diff
@@ -48,69 +67,171 @@ type statelessUnsignedBlock struct {
 	PChainHeight uint64 `serialize:"true"`
 	Certificate  []byte `serialize:"true"`
 	Block        []byte `serialize:"true"`
+	VRFSig       []byte `serialize:"true"`
+}
+
+type unsignedBlockTypes interface {
+	statelessUnsignedBlockV0 | statelessUnsignedBlock
 }
 
-type statelessBlock struct {
-	StatelessBlock statelessUnsignedBlock `serialize:"true"`
-	Signature      []byte                 `serialize:"true"`
+type statelessBlock[T unsignedBlockTypes] struct {
+	StatelessBlock T      `serialize:"true"`
+	Signature      []byte `serialize:"true"`
 
 	id        ids.ID
 	timestamp time.Time
 	cert      *staking.Certificate
 	proposer  ids.NodeID
 	bytes     []byte
+	vrfSig    *bls.Signature
 }
 
-func (b *statelessBlock) ID() ids.ID {
-	return b.id
+func (b *statelessBlock[T]) ParentID() ids.ID {
+	switch sb := any(b.StatelessBlock).(type) {
+	case statelessUnsignedBlock:
+		return sb.ParentID
+	case statelessUnsignedBlockV0:
+		return sb.ParentID
+	default:
+		panic(errUnexpectedStatelessUnsignedBlock)
+	}
+}
+
+func (b *statelessBlock[T]) Block() []byte {
+	switch sb := any(b.StatelessBlock).(type) {
+	case statelessUnsignedBlock:
+		return sb.Block
+	case statelessUnsignedBlockV0:
+		return sb.Block
+	default:
+		panic(errUnexpectedStatelessUnsignedBlock)
+	}
+}
+
+func (b *statelessBlock[T]) VRFSig() []byte {
+	switch sb := any(b.StatelessBlock).(type) {
+	case statelessUnsignedBlock:
+		return sb.VRFSig
+	case statelessUnsignedBlockV0:
+		return nil
+	default:
+		panic(errUnexpectedStatelessUnsignedBlock)
+	}
 }
 
-func (b *statelessBlock) ParentID() ids.ID {
-	return b.StatelessBlock.ParentID
+func (b *statelessBlock[T]) PChainHeight() uint64 {
+	switch sb := any(b.StatelessBlock).(type) {
+	case statelessUnsignedBlock:
+		return sb.PChainHeight
+	case statelessUnsignedBlockV0:
+		return sb.PChainHeight
+	default:
+		panic(errUnexpectedStatelessUnsignedBlock)
+	}
 }
 
-func (b *statelessBlock) Block() []byte {
-	return b.StatelessBlock.Block
+func (b *statelessBlock[T]) innerCertificate() []byte {
+	switch sb := any(b.StatelessBlock).(type) {
+	case statelessUnsignedBlock:
+		return sb.Certificate
+	case statelessUnsignedBlockV0:
+		return sb.Certificate
+	default:
+		panic(errUnexpectedStatelessUnsignedBlock)
+	}
 }
 
-func (b *statelessBlock) Bytes() []byte {
+func (b *statelessBlock[T]) innerTimestamp() int64 {
+	switch sb := any(b.StatelessBlock).(type) {
+	case statelessUnsignedBlock:
+		return sb.Timestamp
+	case statelessUnsignedBlockV0:
+		return sb.Timestamp
+	default:
+		panic(errUnexpectedStatelessUnsignedBlock)
```

Comment:
q: this function seems to be called only in functions returning error. Why not returning the error instead of panicking?
I understand you need to panic in exported methods, but for the unexported ones we may go a different route.
Same for `innerCertificate`

Reply:
these panics would never be invoked ( which is why I used panic.. ).
The underlying datatype must be either `statelessUnsignedBlock` or `statelessUnsignedBlockV0`, and this is validated during compile time by `unsignedBlockTypes`.
but it might not be needed if @dhrubabasu 's approach would work.